File: packages/domain-types/src/asset.ts
import type { CustodyMode } from "./margin.js";

export type AssetStandard = "classic_sac" | "sep57";
export type YieldType = "accumulating" | "distributing";

/** Mirrors the `asset` table (PRD §7). */
export interface AssetConfig {
  id: string;
  code: string;
  issuerG: string;
  /** Issuer registry code (e.g. `"spiko"`, `"etherfuse"`, `"franklin"`) — resolves the
   * `IssuerNavAdapter` in `services/price-guard/src/adapters/registry.ts` by exact match. */
  issuerCode: string;
  contractC: string;
  standard: AssetStandard;
  yieldType: YieldType;
  custodyMode: CustodyMode;
  ccy: string;
  redemptionLagDays: number;
  redemptionDailyCap: bigint;
  navSchedule: string;
  priceBandBps: number;
  dailyMoveBandBps: number;
  haircutBaseBps: number;
  haircutFxBps: number;
  status: string;
}

File: contracts/common/src/asset.rs
//! The asset configuration shape shared by every contract that reads Registry state
//! (`Registry` writes it, `PledgeVault`/`CreditLine`/`ExitDesk`/`RepoDvP` read it via
//! cross-contract calls). Living in `ballast-common` means a cross-contract `asset()` call
//! decodes into the exact same Rust type on both sides — no risk of two independently-defined
//! structs drifting out of sync at the XDR level. Mirrors the `asset` table (PRD §7).

use soroban_sdk::{contracttype, Address, Symbol};

#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum AssetStandard {
    ClassicSac,
    Sep57,
}

#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum YieldType {
    Accumulating,
    Distributing,
}

#[contracttype]
#[derive(Clone, Debug, Eq, PartialEq)]
pub enum CustodyMode {
    Escrow,
    IssuerLien,
    CustodianLien,
}

#[contracttype]
#[derive(Clone, Debug)]
pub struct AssetConfig {
    pub issuer: Address,
    pub contract: Address,
    pub standard: AssetStandard,
    pub yield_type: YieldType,
    pub custody_mode: CustodyMode,
    pub ccy: Symbol,
    pub redemption_lag_days: u32,
    pub haircut_base_bps: u32,
    pub haircut_fx_bps: u32,
    pub price_band_bps: u32,
    pub daily_move_band_bps: u32,
    pub status: Symbol,
}

impl AssetConfig {
    pub fn total_haircut_bps(&self) -> u32 {
        self.haircut_base_bps + self.haircut_fx_bps
    }
}

#[contracttype]
#[derive(Clone, Debug)]
pub enum ParamChange {
    HaircutBaseBps(u32),
    HaircutFxBps(u32),
    PriceBandBps(u32),
    DailyMoveBandBps(u32),
    RedemptionLagDays(u32),
    Status(Symbol),
}
